### PR TITLE
Clean up and simplify contexts as much as possible

### PR DESCRIPTION
**Pull request by imobachgs**
## Tasks

- [x] `LanguageSelector`
- [x] `Storage`
- [x] `Software`

### CODE REVIEW

File: web/src/App.js [hunk 1]
```diff
@@ -1,12 +1,54 @@
...
+  const [isInstalling, setIsInstalling] = useState(false);
+
+  useEffect(autoLogin, []);
+
+  useEffect(() => {
+    if (loggedIn === true) return;
+
+    return client.onSignal('StatusChanged', (_path, _iface, _signal, args) => {
+      setIsInstalling(args[0] !== 0);
+    });
```
Comment (dgdavid) [inline web/src/App.js:42]: NP: we should simplify this.

File: web/src/InstallationProgress.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      setProgress(progress);
+    });
+  }, []);
+
+  return (
+    <Stack hasGutter>
+      <StackItem>
+        <Category title="Progress" icon={ProgressIcon} >
+          <Progress title="Installing" value={Math.round(progress.step / progress.steps * 100)} />
+          <Progress title={progress.title} value={Math.round(progress.substep / progress.substeps * 100)} />
```
Comment (dgdavid) [inline web/src/InstallationProgress.js:50]: Note for future work: show the substeps progress only if there are substeps

File: web/src/LanguageSelector.js [hunk 3]
```diff
@@ -11,49 +11,74 @@ import {
...
+      return { ...state, isFormOpen: false, current: state.initial };
+    }
+
+    case "CHANGE": {
+      return { ...state, current: action.payload };
+    }
+
+    case "OPEN": {
+      return { ...state, isFormOpen: true };
+    }
```
Comment (dgdavid) [inline web/src/LanguageSelector.js:33]: :+1:

File: web/src/LanguageSelector.js [hunk 4]
```diff
@@ -11,49 +11,74 @@ import {
...
+
+export default function LanguageSelector() {
   const client = useInstallerClient();
+  const [state, dispatch] = useReducer(reducer, initialState);
+  const { current: language, languages, isFormOpen } = state;
 
-  useEffect(() => {
-    client.getLanguages().then(setOptions);
+  useEffect(async () => {
+    const languages = await client.getLanguages();
```
Comment (dgdavid) [inline web/src/LanguageSelector.js:51]: Note for the future: to improve the UI feedback, maybe we can add a `loading` property in the state and put it as true here. That way, we can put a spinner somewhere to give feedback to the user while waiting for promises.

File: web/src/LanguageSelector.js [hunk 5]
```diff
@@ -11,49 +11,74 @@ import {
...
-  const onOpen = () => {
-    setLanguage(value);
-    setFormOpen(true);
-  }
+  const open = () => dispatch({ type: "OPEN" });
 
-  const onClose = () => {
-    setFormOpen(false);
+  const cancel = () => {
+    dispatch({ type: "CANCEL" });
```
Comment (dgdavid) [inline web/src/LanguageSelector.js:59]: This one can be written in one line too.

File: web/src/LanguageSelector.js [hunk 6]
```diff
@@ -11,49 +11,74 @@ import {
...
   }
 
-  const applyChanges = () => {
-    onChange(language);
-    onClose();
+  const accept = async () => {
+    // TODO: handle errors
```
Comment (dgdavid) [inline web/src/LanguageSelector.js:63]: The `loading` property mentioned above can be useful here too.

File: web/src/LanguageSelector.js [hunk 7]
```diff
@@ -1,5 +1,5 @@
-import { useState, useEffect } from 'react';
-import { useInstallerClient } from './context/installer';
+import { useReducer, useEffect } from "react";
```
Comment (dgdavid) [inline web/src/LanguageSelector.js:1]: NP: missing copyright header (no related with this PR)

File: web/src/ProductSelector.js [hunk 8]
```diff
@@ -1,4 +1,4 @@
-import { useState, useEffect } from 'react';
```
Comment (dgdavid) [inline web/src/ProductSelector.js:1]: NP: missing copyright header (no related with this PR)

File: web/src/ProductSelector.js [hunk 9]
```diff
@@ -11,38 +11,63 @@ import {
...
-  const onOpen = () => {
-    setProduct(value);
-    setFormOpen(true);
-  }
+  const open = () => dispatch({ type: "OPEN" });
 
-  const onClose = () => {
-    setFormOpen(false);
+  const cancel = () => {
+    dispatch({ type: "CANCEL" });
```
Comment (dgdavid) [inline web/src/ProductSelector.js:59]: NP: can be written in one line too

File: web/src/ProductSelector.js [hunk 10]
```diff
@@ -11,38 +11,63 @@ import {
...
+        return state;
+    }
+  }
+}
+
+const initialState = {
+  products: [], initial: null, current: null, isFormOpen: false
+};
+
+export default function ProductSelector() {
```
Comment (dgdavid) [inline web/src/ProductSelector.js:45]: Note for myself: looking at the similarities of current selectors, looks like could be a good idea to build a `FormSelector` component. But maybe it is too soon yet.

File: web/src/TargetSelector.js [hunk 11]
```diff
@@ -1,4 +1,5 @@
-import { useState } from 'react';
```
Comment (dgdavid) [inline web/src/TargetSelector.js:1]: NP: Missing copyright header (no related with this PR)

File: web/src/lib/InstallerClient.js [hunk 12]
```diff
@@ -96,10 +114,22 @@ export default class InstallerClient {
     )
   }
 
+  async getOption(name) {
+    try {
+      const [{ v: option }] = await this.client().call(
```
Comment (dgdavid) [inline web/src/lib/InstallerClient.js:119]: I love it :stuck_out_tongue_winking_eye: But lets drop a useful link here

* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#nested_object_and_array_destructuring

**Review by dgdavid — APPROVED** [review]
Just a bunch of minor comments, some of them notes for myself (excuse me for that)

LGTM, Thank you!

**Comment by dgdavid** [comment]
> Just a bunch of minor comments, some of them notes for myself (excuse me for that)
> 
> LGTM, Thank you!

Ouch! I didn't notice that it is a draft PR yet :sweat_smile:

File: web/src/LanguageSelector.js (hunk 4 above)
Comment (imobachgs) [reply web/src/LanguageSelector.js:51]: Yes, you are right. We already need to do something like that while deciding whether we should display the login form.

**Review by dgdavid — APPROVED** [review]
LGTM

**Pull request merged**